### ISSUE TITLE
defining an explicit sum of squared error loss function

### ISSUE THREAD

**Issue by stites** [mstksg/backprop#2]
Hey @mstksg -- I was trying to write the above loss function, explictily, and made it to the following:

```haskell

> {-# LANGUAGE ScopedTypeVariables #-}
> {-# LANGUAGE PatternSynonyms #-}
> import qualified Numeric.Backprop as BP
> import qualified Generics.SOP     as SOP
> import Prelude
> import Numeric.LinearAlgebra.Static (L, R, (<.>), konst)
> import GHC.TypeLits (KnownNat)
> import Numeric.Backprop (BPOp, Op, gTuple, (#<~), pattern (:<), (~$))
>
> instance SOP.Generic (R o)
>
> loss :: forall s n .  (KnownNat n) => R n -> BPOp s '[ R n ] Double
> loss target = BP.withInps $ \(outVar :< _) -> do
>   err :< _ <- gTuple #<~ (BP.constVar target - outVar)
>   dot ~$ (err :< BP.only err)

```

using the definition of `dot` from the MNIST tutorial:

```haskell

> dot :: forall n . KnownNat n => Op '[ R n, R n ] Double
> dot = BP.op2' $ \x y ->
>     let
>       back :: Maybe Double -> (R n, R n)
>       back = \case Nothing -> (y, x)
>                    Just g  -> (konst g * y, x * konst g)
>     in
>       (x <.> y, back)

```

But I'm running into a type error where it seems like the type of `R n` is getting lost:

```
QNetworkBackprop.lhs:16:13: error:
    • Couldn't match type ‘hmatrix-0.18.0.0:Internal.Static.Dim
                             n (Data.Vector.Storable.Vector ℝ)’
                     with ‘R n0’
      Expected type: Prod (BVar s '[R n]) '[R n0, R n0]
        Actual type: Prod (BVar s '[R n]) '[a2, a2]
    • In the second argument of ‘(~$)’, namely ‘(err :< err :< Ø)’
      In a stmt of a 'do' block: dot ~$ (err :< err :< Ø)
      In the expression:
        do { err :< _ <- gTuple #<~ (constVar target - outVar);
             dot ~$ (err :< err :< Ø) }
    • Relevant bindings include
        err :: BVar s '[R n] a2 (bound at QNetworkBackprop.lhs:15:5)
        outVar :: BVar s '[R n] a1 (bound at QNetworkBackprop.lhs:14:33)
        target :: R n (bound at QNetworkBackprop.lhs:14:8)
        loss :: R n -> BPOp s '[R n] Double
          (bound at QNetworkBackprop.lhs:14:3)
Failed, modules loaded: none.
```

Would you be able to point me in the right direction? You should be able to copy/paste this issue's markdown into a literate haskell file to see the error. Thanks!

**Comment by stites** [comment]
Okay, I was able to figure it out -- while `dot` is an `Op '[ R n, R n] Double` when lifting to a BPOp with `(~$)`, `dot` operates on values in `BVar s '[R n]`, making the type of the `Prod` it operates on `Prod (BVar s '[R n]) '[R n, R n]`. I thought we could pull out the `R n`s and operate on `Prod _ '[R n, R n]`, but didn't fully understand what functor instance I was using.

Fixing this code with the following, but ignore the commented line until the next paragraph:

```haskell

> loss :: forall s n .  (KnownNat n) => R n -> BPOp s '[ R n ] Double
> loss target = BP.withInps $ \(outVar :< _) -> do
>   -- (err':: BVar s '[R n] (R n)) :< _ <- gTuple #<~ (BP.constVar target - outVar)
>   let
>     err :: BVar s '[R n] (R n)
>     err = BP.constVar target - outVar
>
>   dot ~$ (err :< BP.only err)


```

Unfortunately, I still don't understand the error from earlier. If you uncomment the line, what I now think should be an `err::BVar s '[R n] (R n)`, GHC believes is an unwrapped static hmatrix vector:  `BVar s '[R n] (a::hmatrix-0.18.0.0:Internal.Static.Dim n (Data.Vector.Storable.Vector hmatrix-0.18.0.0:Internal.Static.ℝ))`. Since `R` is just a newtype wrapper around this and backprop is backend-agnostic, shouldn't these types be identical or am I misunderstanding something?

**Comment by mstksg** [comment]
Hm, interesting; can't test this at the moment, but what happens if you annotate `gTuple` with the type you expect?

**Comment by stites** [comment]
I get the following:

```haskell
err' :< _ <- gTuple #<~ (BP.constVar target - out) :: BP.BP s '[R n] (Prod (BVar s '[R n]) '[R n])
```

```
        • Couldn't match type ‘hmatrix-0.18.0.0:Internal.Static.Dim
                                 n (Data.Vector.Storable.Vector LA.ℝ)’
                         with ‘R n’
            arising from a use of ‘gTuple’
        • In the first argument of ‘(#<~)’, namely ‘gTuple’
          In a stmt of a 'do' block:
            err' :< _ <- gTuple #<~ (BP.constVar target - out) ::
                           BP.BP s '[R n] (Prod (BVar s '[R n]) '[R n])
          In the expression:
            do { err' :< _ <- gTuple #<~ (BP.constVar target - out) ::
                                BP.BP s '[R n] (Prod (BVar s '[R n]) '[R n]);
                 let err :: BVar s '[R n] (R n)
                     err = BP.constVar target - out;
                 dot ~$ (err :< only err) }
        • Relevant bindings include
            out :: BVar s '[R n] a1 (bound at src/QNetworkBackprop.hs:201:31)
            target :: R n (bound at src/QNetworkBackprop.hs:201:6)
            loss :: R n -> BPOp s '[R n] Double
```

I'll try to do a little digging myself this weekend.

**Comment by stites** [comment]
With the help of @mitchellwrosen I think I understand what's going on.

The main issue is that `instance SOP.Generic (R n)` unwraps newtype wrappers. `gTuple` is defined as:

```haskell
gTuple :: (SOP.Generic a, SOP.Code a ~ '[as]) => Iso' a (Tuple as)
gTuple = gSOP . sum1
```

and

```haskell
ghci> :kind! SOP.Code (R 1)
SOP.Code (R 1) :: [[ghc-prim-0.5.0.0:GHC.Types.*]]
= '['[hmatrix-0.18.0.0:Internal.Static.Dim
        1 (Data.Vector.Storable.Vector LA.ℝ)]]
```

Although `:kind! SOP.Code (Sum Int)` is a cleaner example.

This isn't really a backprop issue, just a funny interaction between hmatrix and generic-sop. Also, since hmatrix doesn't export the `R` constructor (or `Dim`), there's no bridge between the type- and term- levels to stick `Dim n (LA.Vector ℝ)` back into an `R n` (from what I can tell).

Options around this include the hack of composing `gTuple` with an iso that does `unsafeCoerce`, or just not using gTuple on `R` or `L` (restricting them to ops).

Calling this closed!
